src/core/seat/seat-impl.hpp

```diff
@@ -57,6 +57,7 @@ struct seat_t::impl
 
     void set_keyboard_focus(wf::scene::node_ptr keyboard_focus);
     wf::scene::node_ptr keyboard_focus;
+    bool is_grab = false;
```

Comment:
I am highly sceptical of such approaches as you might have noticed .. what makes grab nodes special here? Why is this not a problem for other nodes (normal client surfaces)?

Reply:
Because it intercepts the client's processing. Consider the following user events:

1. keydown alt
2. keydown tab
3. keyup tab

The switcher plugin activates at event 2. Now the text input is unfocused, so the input releases all pressed keys, sending keyup tab and keyup alt. The latter causes switcher to think the user has released the alt key.

With 49d12183733afeb6f8cf8acb2c6458cd7f56e4f6 the switcher case seems to be fixed, but there may be other complex interactions with other plugins.

Reply:
> so the input releases all pressed keys, sending keyup tab and keyup alt. The latter causes switcher to think the user has released the alt key.

I think I am starting to see what is going on, but why on earth does fcitx5 think it needs to generate these keys? It is not supposed to do so, at least I don't think why it would need that. I'm suspecting that here we're really adding a workaround for an fcitx5 bug.

Reply:
Maybe we could just ignore all IM virtual keyboard events (whether to grab or not) until we switch focus? Just like we ignore commits until the focused text input switches properly. Because like now imagine alt-tab was not a shortcut which opens a grab, but immediately would switch and give focus to the next window. During normal interaction, that would be (this is what an imaginary plugin MIGHT do, I'm not saying we have a plugin which does this atm :)):

1. press alt, press tab
2. old window loses keyboard focus (no need for release events)
3. new window get keyboard focus and is told that alt/tab are pressed on enter
4. user releases alt, tab -> new window gets the release events

So what fcitx5 does would be wrong in this case, because fcitx5 generates additional release events between 2 and 3 (according to your description, if I understood it correctly)

Reply:
> why on earth does fcitx5 think it needs to generate these key releases?

It seems to workaround a sway bug....https://github.com/fcitx/fcitx5/commit/6f92a0cbe7a66924c552cc80d967cc32de9e929e